### PR TITLE
Add verbose messages to tool shims

### PR DESCRIPTION
**Pull request by charlespierce**
Info
-----
* The initial wave of verbose messages only covered the `volta` executable commands, not the behavior of the shims.
* To support diagnostic messages on shims and the `volta list` command, the `Session::current_platform` method should return the source that the platform came from as well as the platform itself.
* This will allow us to avoid duplication of logic checking for `project_platform` and then `user_platform` separately.
* Additionally, when we implement a fix for #436 and have the platforms merged, we will want to make sure that merging logic is centralized in one location and not duplicated.

Changes
-----
* Added `Source`, `SourcedPlatformSpec`, and `SourcedImage` to track the source of a platform / image (`Project` or `User`).
* Cleaned up logic in the tool `command` functions that duplicated the `current_platform` lookup to instead use the source property.
* Added `debug!` messages to all of the tool implementations to include more information about the platform and files that Volta is using.

Screenshots
-----
<details><summary>Node Outside Project</summary>

Node Outside Project
</details>

<details><summary>Node In Project</summary>

Node In Project
</details>

<details><summary>npm Outside Project</summary>

npm Outside Project
</details>

<details><summary>npm In Project</summary>

npm In Project
</details>

<details><summary>Yarn Outside Project</summary>

Yarn Outside Project
</details>

<details><summary>Yarn In Project</summary>

Yarn In Project
</details>

<details><summary>Binary Outside Project</summary>

Binary Outside Project
</details>

<details><summary>Binary In Project</summary>

Binary In Project
</details>

### CODE REVIEW

File: crates/volta-core/src/tool/binary.rs [hunk 1]
```diff
@@ -27,29 +29,38 @@ where
                 });
             }
 
+            debug!("Found bin in project at {}", path_to_bin.display());
```
Comment (rwjblue) [inline crates/volta-core/src/tool/binary.rs]: Should we say "Found a bin for {} in project at {}" where the first substitution is the name of the bin we are looking for? 

All of the messages in this file _might_ want the same treatment?

File: crates/volta-core/src/tool/node.rs [hunk 2]
```diff
@@ -13,11 +15,19 @@ where
...
+                Source::User => debug!("Using node@{} from user default platform", platform.node()),
+            };
+
             let image = platform.checkout(session)?;
             let path = image.path()?;
             Ok(ToolCommand::direct(OsStr::new("node"), args, &path))
         }
-        None => ToolCommand::passthrough(OsStr::new("node"), args, ErrorDetails::NoPlatform),
+        None => {
+            debug!("Could not find platform, delegating to system");
```
Comment (rwjblue) [inline crates/volta-core/src/tool/node.rs]: Probably should include what we are looking for (AFAICT it is `node`, right?).

Comment (charlespierce) [reply crates/volta-core/src/tool/node.rs]: That's a good point, all of the "No platform" messages could be updated to be more specific. Maybe "Could not find volta-managed node, delegating to system"?

File: crates/volta-core/src/tool/binary.rs (hunk 1 above)
Comment (charlespierce) [reply crates/volta-core/src/tool/binary.rs]: That's a good idea, I initially omitted it because the `path_to_bin` will have e.g. `/ember` as the end of it, but I think it probably reads better if the command is included separately.

**Comment by charlespierce** [comment]
@rwjblue I used "platform", because that's the internal name for the block of Node and Optional Yarn that can be specified for running a command. It can come from the user defaults (when you run `volta install node` it sets that platform), from the project itself with the `volta` key in `package.json`, or from the tool settings when you do `volta install ember-cli` (since those are pinned to allow the tools to always run).

We probably don't need to use that language, however, we can use "configuration" or something like that:

```
Using node@10.15.0 from project configuration
Using yarn@1.16.0 from default configuration
Using node@12.1.0 from binary configuration
```

File: crates/volta-core/src/tool/node.rs (hunk 2 above)
Comment (rwjblue) [reply crates/volta-core/src/tool/node.rs]: Ya, that sounds great

File: crates/volta-core/src/tool/binary.rs (hunk 1 above)
Comment (rwjblue) [reply crates/volta-core/src/tool/binary.rs]: Agree, also (not sure if this is true already) we'd want to quote the path and ensure it doesn't get wrapped (wasn't there a util for that already?)

**Comment by rwjblue** [comment]
Ya, `configuration` definitely seems more user friendly than `platform` to me. But I'm also confused about all the different combinations of "configurations". The list above (from the messages in this PR) _seems_ to indicate that there are 4 different platform/configuration types, and AFAICT that is not true.

**Comment by charlespierce** [comment]
I don't believe we have any messages about `system platform` (if we do that's definitely wrong, because that's not a concept we have. The other three are all valid, there's the user defaults, the project-local, and for global tools, we pin the version of `node` that we use to run them (bin platform). That allows the global tools to continue to work, even if you switch default versions of Node, without having to re-install the tool.

File: crates/volta-core/src/tool/binary.rs (hunk 1 above)
Comment (charlespierce) [reply crates/volta-core/src/tool/binary.rs]: There isn't any custom wrapping style applied to the debug output at the moment, so the only wrapping that would happen would be applied by the terminal itself. If / when we add that behavior, we'll definitely use the same logic we already have for the Warning messages that prevents it from wrapping paths on hyphen characters.

File: crates/volta-core/src/session.rs [hunk 3]
```diff
@@ -108,10 +108,13 @@ impl Session {
         self.project.get()
     }
 
-    pub fn current_platform(&self) -> Fallible<Option<Rc<PlatformSpec>>> {
+    pub fn current_platform(&self) -> Fallible<Option<SourcedPlatformSpec>> {
         match self.project_platform()? {
-            Some(platform) => Ok(Some(platform)),
-            None => self.user_platform(),
+            Some(platform) => Ok(Some(SourcedPlatformSpec::project(platform))),
+            None => {
```
Comment (chriskrycho) [inline crates/volta-core/src/session.rs]: *Certainly* not a blocker, but you can simplify this a bit:

```rust
Ok(user.map(SourcedPlatformSpec::user))
```

You can also reduce it into a single expression if you so desire:

```rust
None => self
    .user_platform()
    .map(|platform| platform.map(SourcedPlatformSpec::user)),
```

The form I actually find *clearest*, though, is reworking the whole thing to make the decision tree clearer:

```rust
match (self.project_platform()?, self.user_platform()?) {
    (Some(platform), _) => Ok(Some(SourcedPlatformSpec::project(platform))),
    (_, Some(platform)) => Ok(Some(SourcedPlatformSpec::user(platform))),
    _ => Ok(None),
}
```

File: crates/volta-core/src/tool/node.rs [hunk 4]
```diff
@@ -13,11 +15,21 @@ where
     session.add_event_start(ActivityKind::Node);
 
     match session.current_platform()? {
-        Some(ref platform) => {
+        Some(platform) => {
+            match platform.source() {
+                Source::Project => {
+                    debug!("Using node@{} from project configuration", platform.node())
```
Comment (chriskrycho) [inline crates/volta-core/src/tool/node.rs]: Nit: it'd be nice to use the `style::tool_version` formatter here:

```rust
debug!("Using {} from project configuration", tool_version("node", platform.node()));
```

As an aside, I'd be inclined to do this to avoid having to tweak the messages separately if we reword in the future:

```rust
let source = match platform.source() {
    Source::Project => "project",
    Source::User => "default",
};
let version = tool_version("node", platform.node());
debug!("Using {} from {} configuration", version, source);
```

File: crates/volta-core/src/tool/npm.rs [hunk 5]
```diff
@@ -14,17 +16,28 @@ where
...
                 if let CommandArg::GlobalAdd(package) = check_npm_install() {
                     throw!(ErrorDetails::NoGlobalInstalls { package });
                 }
             }
             let image = platform.checkout(session)?;
             let path = image.path()?;
+
+            match image.source() {
+                Source::Project => {
+                    debug!("Using npm@{} from project configuration", image.node().npm)
```
Comment (chriskrycho) [inline crates/volta-core/src/tool/npm.rs]: Same suggestions as above here.

File: crates/volta-core/src/tool/npx.rs [hunk 6]
```diff
@@ -14,22 +16,34 @@ where
...
             let image = platform.checkout(session)?;
 
             // npx was only included with npm 5.2.0 and higher. If the npm version is less than that, we
             // should include a helpful error message
             let required_npm = VersionSpec::parse_version("5.2.0")?;
-            if image.node.npm >= required_npm {
+            if image.node().npm >= required_npm {
+                match image.source() {
+                    Source::Project => {
+                        debug!("Using npx@{} from project configuration", image.node().npm)
```
Comment (chriskrycho) [inline crates/volta-core/src/tool/npx.rs]: And again!

File: crates/volta-core/src/tool/yarn.rs [hunk 7]
```diff
@@ -16,40 +16,48 @@ where
...
             if intercept_global_installs() {
                 if let CommandArg::GlobalAdd(package) = check_yarn_add() {
                     throw!(ErrorDetails::NoGlobalInstalls { package });
                 }
             }
 
+            // Note: If we've gotten this far, we know there is a yarn version set
+            match platform.source() {
+                Source::Project => debug!(
+                    "Using yarn@{} from project configuration",
```
Comment (chriskrycho) [inline crates/volta-core/src/tool/yarn.rs]: And again. 😀

File: crates/volta-core/src/session.rs (hunk 3 above)
Comment (charlespierce) [reply crates/volta-core/src/session.rs]: I did consider the last of those, and I agree it makes the decision tree clearer, the issue I had with it is that it means the user platform file needs to be read and parsed for every invocation of a tool, even if everything is specified in the Project settings. With the current approach as well as the old code, the user platform isn't looked at unless it's needed.

This may be an early optimization, but especially when adding the verbose messaging, keeping the file access the same seemed important.

I also find the single-expression form slightly harder to grok, with a map within a map, but I like the improvement of using point-free style for the map function.

File: crates/volta-core/src/tool/node.rs (hunk 4 above)
Comment (charlespierce) [reply crates/volta-core/src/tool/node.rs]: Good call in both cases! Updated!

File: crates/volta-core/src/tool/npx.rs (hunk 6 above)
Comment (charlespierce) [reply crates/volta-core/src/tool/npx.rs]: Updated!

File: crates/volta-core/src/tool/yarn.rs (hunk 7 above)
Comment (charlespierce) [reply crates/volta-core/src/tool/yarn.rs]: Updated!

File: crates/volta-core/src/tool/npm.rs (hunk 5 above)
Comment (charlespierce) [reply crates/volta-core/src/tool/npm.rs]: Updated!

File: crates/volta-core/src/session.rs (hunk 3 above)
Comment (chriskrycho) [reply crates/volta-core/src/session.rs]: Totally makes sense. The fact that it's doing a file lookup *does* make a difference here, and I'm inclined to agree with you.

Another alternative I considered here, which is the longest but may be the next-best option to preserve that *and* keep the decision tree equally clear:

```rust
if let Some(platform) = self.project_platform()? {
    Ok(Some(SourcedPlatformSpec::project(platform)))
} else if let Some(platform) = self.user_platform()? {
    Ok(Some(SourcedPlatformSpec::user(platform)))
} else {
    Ok(None)
}
```

At the end of the day, any of these are fine, but I did find the flow a little hard to follow in the original `match`, which is why I mentioned it!

Comment (charlespierce) [reply crates/volta-core/src/session.rs]: That’s a good idea, and makes the changes in the follow-on PR (#470) a little more explicit, so I think I’ll look into that! Thanks!
